widget/editor.go

```diff
@@ -850,21 +1055,34 @@ func (e *Editor) moveWord(distance int) {
 	}
 	for ii := 0; ii < words; ii++ {
 		for r := next(); unicode.IsSpace(r) && !atEnd(); r = next() {
-			e.Move(direction)
+			e.Move(direction, SelectionExtend)
 		}
-		e.Move(direction)
+		e.Move(direction, SelectionExtend)
 		for r := next(); !unicode.IsSpace(r) && !atEnd(); r = next() {
-			e.Move(direction)
+			e.Move(direction, SelectionExtend)
 		}
 	}
+	e.updateSelection(selectionAction)
 }
 
-// deleteWord the next word(s) in the specified direction.
+// deleteWord deletes the next word(s) in the specified direction.
 // Unlike moveWord, deleteWord treats whitespace as a word itself.
 // Positive is forward, negative is backward.
 // Absolute values greater than one will delete that many words.
+//
+// If there is a selection, it is deleted and counts as a single word. As a
+// special case, if distance == 0, deletes the selection (if any) and returns.
```

Comment:
I'd expect distance == 0 to always be a no-op. What's the reason for preferring the selection-clearing behaviour?

Reply:
This particular case in deleteWord was only so that it would have the same behavior as Delete.

In the case of Delete, it was because I wanted to be able to call Delete unconditionally to only delete the selection and not delete anything if there was no selection, in Editor.prepend, and also because as an API user I suspect I'd want that functionality.

In Editor.prepend, I could've make the call conditional on SelectionLen() > 0, but it bugged me that then I'd be doing that comparison twice, once in Editor.prepend, and then again in Delete.

I could make DeleteSelection() that returns true if something was actually deleted and use that in Editor.prepend and also in deleteWord, and then Delete and deleteWord could both be complete no-ops for 0.

Reply:
> This particular case in deleteWord was only so that it would have the same behavior as Delete.
> 
> In the case of Delete, it was because I wanted to be able to call Delete unconditionally to only delete the selection and not delete anything if there was no selection, in Editor.prepend, and also because as an API user I suspect I'd want that functionality.
> 

If you don't have a clear external use yet, I'm inclined towards the "0 means no-op" behaviour. What I imagine someone doing is something like

var deleteCounter int
var e Editor
for ... {
   ...
   deleteCounter++
   ...
}
e.Delete(deleteCounter)

This code works as expected for the usual case (no selection), but subtly fails when something is selected.

> In Editor.prepend, I could've make the call conditional on SelectionLen() > 0, but it bugged me that then I'd be doing that comparison twice, once in Editor.prepend, and then again in Delete.
> 

That sounds more like prepend/append should be merged, inlined or something else. FWIW, append/prepend are not my favorites either.

> I could make DeleteSelection() that returns true if something was actually deleted and use that in Editor.prepend and also in deleteWord, and then Delete and deleteWord could both be complete no-ops for 0.

DeleteSelection may be needed, but how often? I'd say a e.DeleteRange(e.Selection()) is a better way to expose your Delete(0). Returning a bool seems ugly, though. If you're interested in the result, why not check SelectionLen first?

Reply:
> If you don't have a clear external use yet, I'm inclined towards the "0 means no-op" behaviour

Okay, I'll change it.

Reply:
Refactored this extensively.  PTAL.